core/src/main/java/org/apache/accumulo/core/file/blockfile/cache/lru/LruBlockCache.java

```diff
@@ -410,14 +411,15 @@ public void add(CachedBlock block) {
     }
 
     public long free(long toFree) {
-      CachedBlock[] blocks = queue.get();
+      Collection<CachedBlock> blocks = queue.getAll();
```

Comment:
This won't work. The iterator doesn't preserve the queue ordering. This is the same reason why the `toArray()` needed to be subsequently sorted. Rather than use the iterator, it needs to use the `poll()` to grab items off the queue in order. Something like:

```java
CachedBlock block = null;
while ((block = queue.poll()) != null) {
  // ...
}
```

We need to evict in order, otherwise, we're not properly freelying LFU blocks.

Reply:
Hey @ctubbsii thanks for being so hands on here.  Appreciate it.

Ya, I really screwed up on this one.  I think I was doing too much coding yesterday.  Thanks for the double-check.

I had it in my head that this `free` method was `freeAll`, in which case order doesn't matter.  Let me take another stab at that.

But now that I look at it, I think `trunk` is pretty broken.  We know that the `get()` call on the queue creates a list of all the blocks in the queue by removing them in order.  However, the `free(long toFree)` method will only 'free' up to the amount of `toFree`.  The rest of the blocks in the list are lost.

Reply:
OK.  I've been looking at this for a few minutes.  I understand it better now.  There is no leak here... the code is very confusing.  The purpose of these classes to find find the Top N `CachedBlocks` where N is LRU, but it also the `sum(N)` is less than some threshold.  Nothing is being leaked here.  The code becomes a lot easier if we can relax the `sum(N)` part and just order all of the blocks by access the peel off the top N.   Not sure if that's the best way to go though, need to think about it some more.

Reply:
> Hey @ctubbsii thanks for being so hands on here. Appreciate it.

No worries. Before you issued this PR, I was entirely unfamiliar with this part of Accumulo's code, and it's taking me some time to understand it, but I'm happy to do so to help anybody interested in cleaning up our technical debt.

> But now that I look at it, I think trunk is pretty broken. ... The rest of the blocks in the list are lost.

I'm not yet convinced it's "broken"... but certainly the code that's there now is very messy and confusing. This queue, from which the remaining blocks are "lost", appears to be temporary, used only to organize the blocks to determine priority for eviction during the eviction thread, so it's fine that they are "lost". Being dropped from this data structure (or the entire queue being dropped at the end of the eviction cycle) isn't important, because it's not the same as having `LruBlockCache.evictBlock(block)` called on them. Those blocks that weren't evicted will remain in the cache in `LruBlockCache.map`.

Reply:
@ctubbsii  OK, so I think the `CachedBlockQueue` was trying to be a little bit too cute.  I relaxed the restriction that the queue "consume as close to the specified maxSize as possible."  That is a tricky prospect and it wasn't quite correct in the existing implementation, I think the existing unit tests are wrong so it wasn't caught.

As you can see from my added unit tests, I had to manipulate the `CachedBlock` `recordSize` not the just set the 'blockSize' in the constructor to make the queue function properly.  The existing unit tests do not do this step, so every block will be added to the queue because the `heapSize` of the queue is always zero.

Reply:
Was being bit by another issue:

```
    public long free(long toFree) {
      CachedBlock[] blocks = queue.get();
      long freedBytes = 0;
      for (CachedBlock block : blocks) {
        freedBytes += evictBlock(block);
        if (freedBytes >= toFree) {
          return freedBytes;
        }
      }
      return freedBytes;
    }
```

This method will always evict the first block... even if... for example... `toFree` is a negative number.  In my implementation, if `toFree` is a negative number, then it is treated as a no-op.  Not sure what the better behavior is.  I see in the unit test, a negative number is being passed in.

Reply:
Yup.  I figured it out.  The unit test was incorrect and therefore it was hiding this issue where it would always evict the first block, even if it wasn't required to do so.

I've updated the unit test so now my implementation works (where it won't evict if it doesn't have to)

Reply:
This code is never called if the number was negative. That's checked in `LruBlockCache.evict()` before we get here. By the time this code runs, we *always* want to evict at least one block.